Utils/src/HiddenSectorProducer.cc

```diff
@@ -22,143 +27,390 @@
+  iEvent.getByToken(GenJetTok_, h_genjets);
+
+  auto jet_isHV_vec = std::make_unique<std::vector<bool>>();
+  auto hvCategory = std::make_unique<std::vector<int>>();
+  auto darkPtFrac = std::make_unique<std::vector<double>>();
+  auto MT2JetsID = std::make_unique<std::vector<int>>();
+  auto genIndex = std::make_unique<std::vector<int>>();
+
+  LorentzVector vpartsSum;
+  if(h_parts.isValid()){
+    for(const auto& i_part : *(h_parts.product())){
+      if(isDark(DarkStableIDs_,i_part.pdgId())){
+        vpartsSum += i_part.p4();
+      }
+    }
+  }
+
+  double MJJ = 0, Mmc = 0, MT = 0, tchMT2 = 0, DeltaPhi1 = 10, DeltaPhi2 = 10, DeltaPhiMin = 10;
+  //only fill these parts if there are >=2 jets
+  if(h_jets->size()>=2){
+    //delta phi
+    DeltaPhi1 = std::abs(reco::deltaPhi(h_jets->at(0).phi(),h_mets->at(0).phi()));
+    DeltaPhi2 = std::abs(reco::deltaPhi(h_jets->at(1).phi(),h_mets->at(0).phi()));
+    DeltaPhiMin = std::min(DeltaPhi1,DeltaPhi2);
+
+    //masses
+    LorentzVector vjj = h_jets->at(0).p4()+h_jets->at(1).p4();
+    MJJ = vjj.M();
+
+    //include all jets in MC mass
+    LorentzVector vmc = vjj + vpartsSum;
+    Mmc = vmc.M();
+
+    //MET is massless, but jets aren't
+    double MET = h_mets->at(0).pt(), METPhi = h_mets->at(0).phi();
+    MT = TransverseMass(vjj.Px(),vjj.Py(),vjj.M(),MET*std::cos(METPhi),MET*std::sin(METPhi),0);
+  }
+
+  //ISRJetProducer method, find jets with hard process particles at their 'core' and other jets are ISR
+  if(h_parts.isValid()){
+    for(const auto& i_jet : *(h_jets.product())){ // loop over AK8 jets
+      bool matched = false; // matched == true means that this jet has a hard process (descendant particle of the Z') at its 'core'
+      for (const auto& i_part : *(h_parts.product())){ // loop over GenParticles
+        if (matched) break; // only need to match one particle to the jet to tag it as FSR
+        if(i_part.status()!=23) continue; // only want particles outgoing from the hard process
+        if(isDark(DarkMediatorIDs_,i_part.mother())) continue; // only want direct descendants of Z' (kind of redundant)
+        //check against daughters in case of hard initial splitting, from ISRJetProducer...
+        std::vector<CandPtr> listOfDaughters;
+        addDaughters(&i_part,listOfDaughters);
+        for (const auto& daughter : listOfDaughters) {
+          float dR = deltaR(i_jet, daughter->p4());
+          if(dR<0.8){
+            matched = true;
+            break;
+          }
+        }
+      }
+      jet_isHV_vec->push_back(matched);
+    }
+
+    //t-channel jet categorization
+    //gen particle organization: last daughters of last copies of dark hadrons, first copies of dark quarks/gluons/mediators, first copies of quarks/gluons/SM quarks from mediator
+    //naming scheme: dark particles Pd, SM particles Ps; P = D (generic daughters), Q (quarks), G (gluons), M (mediators)
+    CandPtr firstQdM1, firstQdM2, firstQsM1, firstQsM2;
```

Comment:
as noted above, try to organize `firstDark` objects vs `medDecay` objects separately to make things clearer